### PR TITLE
cmake: out of source support

### PR DESCRIPTION
**Pull request by elms**
# Description

Add option (`WOLFSSL_BUILD_OUT_OF_TREE`) to make avoid making changes to the source tree. Defaults to value of `CMAKE_DISABLE_SOURCE_CHANGES` if is set.

# Testing

Out of tree builds with options enabled.
Also tested with `vcpkg` configuration which enables `CMAKE_DISABLE_SOURCE_CHANGES`

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1683,7 +1683,22 @@ add_definitions("-DWOLFSSL_IGNORE_FILE_WARN")
...
 # Generate user options header
 message("Generating user options header...")
-set(OPTION_FILE "${CMAKE_CURRENT_SOURCE_DIR}/wolfssl/options.h")
+if (${CMAKE_DISABLE_SOURCE_CHANGES})
+    set(WOLFSSL_BUILD_OUT_OF_TREE_DEFAULT "${CMAKE_DISABLE_SOURCE_CHANGES}")
+else()
+    set(WOLFSSL_BUILD_OUT_OF_TREE_DEFAULT "no")
+endif()
+add_option("WOLFSSL_BUILD_OUT_OF_TREE"
+    "Don't generate files in the source tree (default: ${CMAKE_DISABLE_SOURCE_CHANGES})"
```
Comment (haydenroche5) [inline CMakeLists.txt:1692]: I think this is functionally the same, but I would change `CMAKE_DISABLE_SOURCE_CHANGES` here to `WOLFSSL_BUILD_OUT_OF_TREE_DEFAULT`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2034,10 +2048,20 @@ set(INSTALLED_EXAMPLES
...
+        )
 # Install the headers
+install(DIRECTORY ${WOLFSSL_OUTPUT_BASE}/wolfssl/
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/wolfssl
+        FILES_MATCHING PATTERN "*.h"
+        REGEX ${EXCLUDED_HEADERS_REGEX} EXCLUDE)
+install(DIRECTORY ${WOLFSSL_OUTPUT_BASE}/cyassl/
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/cyassl
+        FILES_MATCHING PATTERN "*.h"
+        REGEX ${EXCLUDED_HEADERS_REGEX} EXCLUDE)
```
Comment (haydenroche5) [inline CMakeLists.txt:2064]: Thanks for the cleanups on the install pieces.

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (elms) [reply CMakeLists.txt]: Done.

**Comment by elms** [comment]
@haydenroche5 rebased and added a few spelling fixes in comments.

**Review by dgarske — APPROVED** [review]

**Review by haydenroche5 — APPROVED** [review]

**Pull request merged**